File: app/src/main/java/com/kickstarter/viewmodels/ProfileViewModel.java
package com.kickstarter.viewmodels;

import android.support.annotation.NonNull;
import android.util.Pair;

import com.kickstarter.libs.ActivityViewModel;
import com.kickstarter.libs.ApiPaginator;
import com.kickstarter.libs.CurrentUserType;
import com.kickstarter.libs.Environment;
import com.kickstarter.libs.rx.transformers.Transformers;
import com.kickstarter.libs.utils.IntegerUtils;
import com.kickstarter.libs.utils.NumberUtils;
import com.kickstarter.models.Project;
import com.kickstarter.models.User;
import com.kickstarter.services.ApiClientType;
import com.kickstarter.services.DiscoveryParams;
import com.kickstarter.services.apiresponses.DiscoverEnvelope;
import com.kickstarter.ui.activities.ProfileActivity;
import com.kickstarter.ui.adapters.ProfileAdapter;
import com.kickstarter.ui.viewholders.EmptyProfileViewHolder;
import com.kickstarter.ui.viewholders.ProfileCardViewHolder;

import java.util.List;

import rx.Observable;
import rx.subjects.PublishSubject;

public interface ProfileViewModel {

  interface Inputs {
    /** Call when the Explore Projects button in the empty state has been clicked. */
    void exploreProjectsButtonClicked();

    /** Call when the messages button has been clicked. */
    void messsagesButtonClicked();

    /** Call when the next page has been invoked. */
    void nextPage();

    /** Call when a project card has been clicked. */
    void projectCardClicked(final @NonNull Project project);
  }

  interface Outputs {
    /** Emits the user avatar image to be displayed. */
    Observable<String> avatarImageViewUrl();

    /** Emits when the backed projects count should be hidden. */
    Observable<Boolean> backedCountTextViewHidden();

    /** Emits the backed projects count to be displayed. */
    Observable<String> backedCountTextViewText();

    /** Emits when the backed projects text view should be hidden. */
    Observable<Boolean> backedTextViewHidden();

    /** Emits when the created projects count should be hidden. */
    Observable<Boolean> createdCountTextViewHidden();

    /** Emits the created projects count to be displayed. */
    Observable<String> createdCountTextViewText();

    /** Emits when the created projects text view should be hidden. */
    Observable<Boolean> createdTextViewHidden();

    /** Emits when the divider view should be hidden. */
    Observable<Boolean> dividerViewHidden();

    /** Emits a list of projects to display in the profile. */
    Observable<List<Project>> projects();

    /** Emits when we should start the {@link com.kickstarter.ui.activities.MessageThreadsActivity}. */
    Observable<Void> startMessageThreadsActivity();

    /** Emits when we should start the {@link com.kickstarter.ui.activities.ProjectActivity}. */
    Observable<Project> startProjectActivity();

    /** Emits when we should resume the {@link com.kickstarter.ui.activities.DiscoveryActivity}. */
    Observable<Void> resumeDiscoveryActivity();

    /** Emits the user name to be displayed. */
    Observable<String> userNameTextViewText();
  }

  final class ViewModel extends ActivityViewModel<ProfileActivity> implements ProfileAdapter.Delegate, Inputs, Outputs {
    private final ApiClientType client;
    private final CurrentUserType currentUser;

    public ViewModel(final @NonNull Environment environment) {
      super(environment);

      this.client = environment.apiClient();
      this.currentUser = environment.currentUser();

      final Observable<User> freshUser = this.client.fetchCurrentUser()
        .retry(2)
        .compose(Transformers.neverError());
      freshUser.subscribe(this.currentUser::refresh);

      final DiscoveryParams params = DiscoveryParams.builder()
        .backed(1)
        .perPage(18)
        .sort(DiscoveryParams.Sort.ENDING_SOON)
        .build();

      final ApiPaginator<Project, DiscoverEnvelope, DiscoveryParams> paginator =
        ApiPaginator.<Project, DiscoverEnvelope, DiscoveryParams>builder()
          .nextPage(this.nextPage)
          .envelopeToListOfData(DiscoverEnvelope::projects)
          .envelopeToMoreUrl(env -> env.urls().api().moreProjects())
          .loadWithParams(__ -> this.client.fetchProjects(params))
          .loadWithPaginationPath(this.client::fetchProjects)
          .build();

      final Observable<User> loggedInUser = this.currentUser.loggedInUser();

      this.avatarImageViewUrl = loggedInUser.map(u -> u.avatar().medium());

      this.backedCountTextViewHidden = loggedInUser
        .map(u -> IntegerUtils.isZero(u.backedProjectsCount()));
      this.backedTextViewHidden = this.backedCountTextViewHidden;

      this.backedCountTextViewText = loggedInUser
        .map(User::backedProjectsCount)
        .filter(IntegerUtils::isNonZero)
        .map(NumberUtils::format);

      this.createdCountTextViewHidden = loggedInUser
        .map(u -> IntegerUtils.isZero(u.createdProjectsCount()));
      this.createdTextViewHidden = this.createdCountTextViewHidden;

      this.createdCountTextViewText = loggedInUser
        .map(User::createdProjectsCount)
        .filter(IntegerUtils::isNonZero)
        .map(NumberUtils::format);

      this.dividerViewHidden = Observable.combineLatest(
        this.backedTextViewHidden,
        this.createdTextViewHidden,
        Pair::create
      )
        .map(p -> p.first || p.second);

      this.projects = paginator.paginatedData();
      this.resumeDiscoveryActivity = this.exploreProjectsButtonClicked;
      this.startProjectActivity = this.projectCardClicked;
      this.startMessageThreadsActivity = this.messsagesButtonClicked;
      this.userNameTextViewText = loggedInUser.map(User::name);

      this.koala.trackProfileView();
    }

    private final PublishSubject<Void> exploreProjectsButtonClicked = PublishSubject.create();
    private final PublishSubject<Void> messsagesButtonClicked = PublishSubject.create();
    private final PublishSubject<Void> nextPage = PublishSubject.create();
    private final PublishSubject<Project> projectCardClicked = PublishSubject.create();

    private final Observable<String> avatarImageViewUrl;
    private final Observable<Boolean> backedCountTextViewHidden;
    private final Observable<String> backedCountTextViewText;
    private final Observable<Boolean> backedTextViewHidden;
    private final Observable<Boolean> createdCountTextViewHidden;
    private final Observable<String> createdCountTextViewText;
    private final Observable<Boolean> createdTextViewHidden;
    private final Observable<Boolean> dividerViewHidden;
    private final Observable<List<Project>> projects;
    private final Observable<Void> resumeDiscoveryActivity;
    private final Observable<Project> startProjectActivity;
    private final Observable<Void> startMessageThreadsActivity;
    private final Observable<String> userNameTextViewText;

    public final ProfileViewModel.Inputs inputs = this;
    public final ProfileViewModel.Outputs outputs = this;

    @Override public void emptyProfileViewHolderExploreProjectsClicked(final @NonNull EmptyProfileViewHolder viewHolder) {
      this.exploreProjectsButtonClicked();
    }
    @Override public void exploreProjectsButtonClicked() {
      this.exploreProjectsButtonClicked.onNext(null);
    }
    @Override public void messsagesButtonClicked() {
      this.messsagesButtonClicked.onNext(null);
    }
    @Override public void nextPage() {
      this.nextPage.onNext(null);
    }
    @Override public void profileCardViewHolderClicked(final @NonNull ProfileCardViewHolder viewHolder, final @NonNull Project project) {
      this.projectCardClicked(project);
    }
    @Override public void projectCardClicked(final @NonNull Project project) {
      this.projectCardClicked.onNext(project);
    }

    @Override public @NonNull Observable<String> avatarImageViewUrl() {
      return this.avatarImageViewUrl;
    }
    @Override public @NonNull Observable<String> backedCountTextViewText() {
      return this.backedCountTextViewText;
    }
    @Override public @NonNull Observable<Boolean> backedCountTextViewHidden() {
      return this.backedCountTextViewHidden;
    }
    @Override public @NonNull Observable<Boolean> backedTextViewHidden() {
      return this.backedTextViewHidden;
    }
    @Override public @NonNull Observable<Boolean> createdCountTextViewHidden() {
      return this.createdCountTextViewHidden;
    }
    @Override public @NonNull Observable<String> createdCountTextViewText() {
      return this.createdCountTextViewText;
    }
    @Override public @NonNull Observable<Boolean> createdTextViewHidden() {
      return this.createdTextViewHidden;
    }
    @Override public @NonNull Observable<Boolean> dividerViewHidden() {
      return this.dividerViewHidden;
    }
    @Override public @NonNull Observable<List<Project>> projects() {
      return this.projects;
    }
    @Override public @NonNull Observable<Void> resumeDiscoveryActivity() {
      return this.resumeDiscoveryActivity;
    }
    @Override public @NonNull Observable<Project> startProjectActivity() {
      return this.startProjectActivity;
    }
    @Override public @NonNull Observable<Void> startMessageThreadsActivity() {
      return this.startMessageThreadsActivity;
    }
    @Override public @NonNull Observable<String> userNameTextViewText() {
      return this.userNameTextViewText;
    }
  }
}

File: app/src/main/java/com/kickstarter/viewmodels/ProfileViewModel.kt
package com.kickstarter.viewmodels

import android.util.Pair
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import com.kickstarter.libs.ApiPaginatorV2
import com.kickstarter.libs.Environment
import com.kickstarter.libs.rx.transformers.Transformers.neverErrorV2
import com.kickstarter.libs.utils.EventContextValues
import com.kickstarter.libs.utils.NumberUtils
import com.kickstarter.libs.utils.extensions.addToDisposable
import com.kickstarter.libs.utils.extensions.isNonZero
import com.kickstarter.libs.utils.extensions.isNotNull
import com.kickstarter.libs.utils.extensions.isZero
import com.kickstarter.models.Project
import com.kickstarter.services.DiscoveryParams
import com.kickstarter.services.apiresponses.DiscoverEnvelope
import com.kickstarter.ui.adapters.ProfileAdapter
import com.kickstarter.ui.viewholders.EmptyProfileViewHolder
import com.kickstarter.ui.viewholders.ProfileCardViewHolder
import io.reactivex.Observable
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.subjects.BehaviorSubject
import io.reactivex.subjects.PublishSubject

interface ProfileViewModel {

    interface Inputs {
        /** Call when the Explore Projects button in the empty state has been clicked.  */
        fun exploreProjectsButtonClicked()

        /** Call when the messages button has been clicked.  */
        fun messagesButtonClicked()

        /** Call when the next page has been invoked.  */
        fun nextPage()

        /** Call when a project card has been clicked.  */
        fun projectCardClicked(project: Project)
    }

    interface Outputs {
        /** Emits the user avatar image to be displayed.  */
        fun avatarImageViewUrl(): Observable<String>

        /** Emits when the backed projects count should be hidden.  */
        fun backedCountTextViewHidden(): Observable<Boolean>

        /** Emits the backed projects count to be displayed.  */
        fun backedCountTextViewText(): Observable<String>

        /** Emits when the backed projects text view should be hidden.  */
        fun backedTextViewHidden(): Observable<Boolean>

        /** Emits when the created projects count should be hidden.  */
        fun createdCountTextViewHidden(): Observable<Boolean>

        /** Emits the created projects count to be displayed.  */
        fun createdCountTextViewText(): Observable<String>

        /** Emits when the created projects text view should be hidden.  */
        fun createdTextViewHidden(): Observable<Boolean>

        /** Emits when the divider view should be hidden.  */
        fun dividerViewHidden(): Observable<Boolean>

        /** Emits a boolean indicating whether projects are being fetched from the API.  */
        fun isFetchingProjects(): Observable<Boolean>

        /** Emits a list of projects to display in the profile.  */
        fun projectList(): Observable<List<Project>>

        /** Emits when we should resume the [com.kickstarter.ui.activities.DiscoveryActivity].  */
        fun resumeDiscoveryActivity(): Observable<Unit>

        /** Emits when we should start the [com.kickstarter.ui.activities.MessageThreadsActivity].  */
        fun startMessageThreadsActivity(): Observable<Unit>

        /** Emits when we should start the [com.kickstarter.ui.activities.ProjectActivity].  */
        fun startProjectActivity(): Observable<Project>

        /** Emits the user name to be displayed.  */
        fun userNameTextViewText(): Observable<String>
    }

    class ProfileViewModel(environment: Environment) : ViewModel(), ProfileAdapter.Delegate, Inputs, Outputs {
        private val client = requireNotNull(environment.apiClientV2())
        private val currentUser = requireNotNull(environment.currentUserV2())
        private val analytics = requireNotNull(environment.analytics())

        private val exploreProjectsButtonClicked = PublishSubject.create<Unit>()
        private val messagesButtonClicked = PublishSubject.create<Unit>()
        private val nextPage = PublishSubject.create<Unit>()
        private val refresh = PublishSubject.create<DiscoveryParams>()
        private val projectCardClicked = PublishSubject.create<Project>()

        private val avatarImageViewUrl: Observable<String>
        private val backedCountTextViewHidden: Observable<Boolean>
        private val backedCountTextViewText: Observable<String>
        private val backedTextViewHidden: Observable<Boolean>
        private val createdCountTextViewHidden: Observable<Boolean>
        private val createdCountTextViewText: Observable<String>
        private val createdTextViewHidden: Observable<Boolean>
        private val dividerViewHidden: Observable<Boolean>
        private val isFetchingProjects = BehaviorSubject.create<Boolean>()
        private val projectList: Observable<List<Project>>
        private val resumeDiscoveryActivity: Observable<Unit>
        private val startMessageThreadsActivity: Observable<Unit>
        private val userNameTextViewText: Observable<String>

        val inputs: Inputs = this
        val outputs: Outputs = this

        val disposables = CompositeDisposable()

        init {
            val freshUser = this.client.fetchCurrentUser()
                .retry(2)
                .compose(neverErrorV2())

            freshUser.subscribe { this.currentUser.refresh(it) }
                .addToDisposable(disposables)

            val params = DiscoveryParams.builder()
                .backed(1)
                .perPage(18)
                .sort(DiscoveryParams.Sort.ENDING_SOON)
                .build()

            val paginator = ApiPaginatorV2.builder<Project, DiscoverEnvelope, DiscoveryParams>()
                .nextPage(nextPage)
                .startOverWith(Observable.just(params))
                .envelopeToListOfData { it.projects() }
                .envelopeToMoreUrl { env -> env.urls()?.api()?.moreProjects() }
                .loadWithParams { this.client.fetchProjects(params).compose(neverErrorV2()) }
                .loadWithPaginationPath { this.client.fetchProjects(it).compose(neverErrorV2()) }
                .build()

            paginator.isFetching
                .subscribe { this.isFetchingProjects.onNext(it) }
                .addToDisposable(disposables)

            val loggedInUser = this.currentUser.loggedInUser()

            this.avatarImageViewUrl = loggedInUser.map { u -> u.avatar().medium() }

            this.backedCountTextViewHidden = loggedInUser
                .map { u -> u.backedProjectsCount().isZero() }
            this.backedTextViewHidden = this.backedCountTextViewHidden

            this.backedCountTextViewText = loggedInUser
                .map<Int> { it.backedProjectsCount() }
                .filter { it.isNonZero() }
                .map { NumberUtils.format(it) }

            this.createdCountTextViewHidden = loggedInUser
                .map { u -> u.createdProjectsCount().isZero() }
            this.createdTextViewHidden = this.createdCountTextViewHidden

            this.createdCountTextViewText = loggedInUser
                .map<Int> { it.createdProjectsCount() }
                .filter { it.isNonZero() }
                .map { NumberUtils.format(it) }

            this.dividerViewHidden = Observable.combineLatest<Boolean, Boolean, Pair<Boolean, Boolean>>(
                this.backedTextViewHidden,
                this.createdTextViewHidden
            ) { a, b -> Pair.create(a, b) }
                .map { p -> p.first || p.second }

            this.projectList = paginator.paginatedData().filter { it.isNotNull() }
            this.resumeDiscoveryActivity = this.exploreProjectsButtonClicked

            this.startMessageThreadsActivity = this.messagesButtonClicked
            this.userNameTextViewText = loggedInUser.map { it.name() }

            projectCardClicked
                .subscribe { analytics.trackProjectCardClicked(it, EventContextValues.ContextPageName.PROFILE.contextName) }
                .addToDisposable(disposables)
        }

        override fun emptyProfileViewHolderExploreProjectsClicked(viewHolder: EmptyProfileViewHolder) = this.exploreProjectsButtonClicked()

        override fun exploreProjectsButtonClicked() = this.exploreProjectsButtonClicked.onNext(Unit)

        override fun messagesButtonClicked() = this.messagesButtonClicked.onNext(Unit)

        override fun nextPage() = this.nextPage.onNext(Unit)

        override fun profileCardViewHolderClicked(viewHolder: ProfileCardViewHolder, project: Project) = this.projectCardClicked(project)

        override fun projectCardClicked(project: Project) = this.projectCardClicked.onNext(project)

        override fun avatarImageViewUrl() = this.avatarImageViewUrl

        override fun backedCountTextViewText() = this.backedCountTextViewText

        override fun backedCountTextViewHidden() = this.backedCountTextViewHidden

        override fun backedTextViewHidden() = this.backedTextViewHidden

        override fun createdCountTextViewHidden() = this.createdCountTextViewHidden

        override fun createdCountTextViewText() = this.createdCountTextViewText

        override fun createdTextViewHidden() = this.createdTextViewHidden

        override fun dividerViewHidden() = this.dividerViewHidden

        override fun isFetchingProjects(): BehaviorSubject<Boolean> = this.isFetchingProjects

        override fun projectList() = this.projectList

        override fun resumeDiscoveryActivity() = this.resumeDiscoveryActivity

        override fun startProjectActivity() = this.projectCardClicked

        override fun startMessageThreadsActivity() = this.startMessageThreadsActivity

        override fun userNameTextViewText() = this.userNameTextViewText

        override fun onCleared() {
            disposables.clear()
            super.onCleared()
        }
    }

    class Factory(private val environment: Environment) :
        ViewModelProvider.Factory {
        override fun <T : ViewModel> create(modelClass: Class<T>): T {
            return ProfileViewModel(environment) as T
        }
    }
}
